module-system/sov-modules-core/src/storage/scratchpad.rs

```diff
@@ -460,6 +479,69 @@ impl<'a, C: Context> StateReaderAndWriter for AccessoryWorkingSet<'a, C> {
     }
 }
 
+/// Module for archival state
+pub mod archival_state {
+    use super::*;
+
+    /// Enum for tagging the specific type of storage.
+    enum ArchivalStore<S: Storage> {
+        /// JMT store
+        Jmt(S),
+        /// Accessory store
+        Accessory(S),
+    }
+
+    /// Archival working set
+    pub struct ArchivalWorkingSet<C: Context> {
+        delta: ArchivalStore<<C as Spec>::Storage>,
+        archival_version: Version,
+        witness: <<C as Spec>::Storage as Storage>::Witness,
+    }
+
+    impl<C: Context> ArchivalWorkingSet<C> {
+        /// Creates a new [`ArchivalWorkingSet`] instance backed by the given [`Storage`]. and [`Version`]
+        /// For jmt state access
+        pub fn new(inner: &<C as Spec>::Storage, version: Version) -> Self {
+            Self {
+                delta: ArchivalStore::Jmt(inner.clone()),
+                archival_version: version,
+                witness: Default::default(),
+            }
+        }
+
+        /// Creates a new [`ArchivalWorkingSet`] instance backed by the given [`Storage`]. and [`Version`]
+        /// For accestory state access
+        pub fn new_accessory(inner: &<C as Spec>::Storage, version: Version) -> Self {
+            Self {
+                delta: ArchivalStore::Accessory(inner.clone()),
+                archival_version: version,
+                witness: Default::default(),
+            }
+        }
+    }
+
+    impl<C: Context> StateReaderAndWriter for ArchivalWorkingSet<C> {
+        fn get(&mut self, key: &StorageKey) -> Option<StorageValue> {
+            match &self.delta {
+                ArchivalStore::Jmt(jmt_store) => {
+                    jmt_store.get(key, Some(self.archival_version), &self.witness)
+                }
+                ArchivalStore::Accessory(accessory_store) => {
+                    if !cfg!(feature = "native") {
```

Comment:
I think this should guard the entire function and we should panic in this case.

Reply:
Here is the other place that needs to be changed: https://github.com/Sovereign-Labs/sovereign-sdk/blob/nightly/module-system/sov-modules-core/src/storage/mod.rs#L209